#**ESCUELA POLITECNICA NACIONAL**
##RECUPERACION INFORMACION
###Examen 01
 Sergio Guaman

En esta parte iniciamops descargando con El comando pip install scikit-learn  la cual instala la librería scikit-learn, que es una herramienta para realizar tareas de aprendizaje automático (machine learning) como clasificación, regresión y clustering.
además instalamos nltk y spacy para la lematización


In [3]:
pip install nltk


In [4]:
pip install scikit-learn

In [5]:
pip install spacy

In [6]:
import os
import re
import nltk

import time # Evaluar tecnicas de vectorización
import psutil # Evaluar tecnicas de vectorización
import gc # Evaluar tecnicas de vectorización
import numpy as np
import pandas as pd # Obtener tablas
import matplotlib.pyplot as plt # Para obtener gráfica
import seaborn as sns # Para obtener gráfica
from collections import defaultdict # Para el indice invertido
from sklearn.preprocessing import normalize # para Word2Vec
from nltk.tokenize import word_tokenize # para tokenizar
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from gensim.models import Word2Vec
from sklearn.metrics.pairwise import cosine_similarity
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

In [7]:
#import spacy

#Obtenemos la base de datos

Lo que se decidió hacer es crear 2 df de rview y movies el cual seguido convertimos en df, luego unimos ambas en uno solo df llamado df_total, con un merged tomando en base a rotten_tomatoes_link

In [8]:
df_review=pd.read_csv("rotten_tomatoes_critic_reviews.csv",sep=',')

In [9]:
df_movies=pd.read_csv("rotten_tomatoes_movies.csv",sep=',')

In [10]:
documentos = {}

In [11]:

df_total = pd.merge(df_review, df_movies, on='rotten_tomatoes_link', how='inner')

In [12]:
df_total=df_total.drop(['original_release_date'],axis=1)


Realizamos un drop de las columnas que no vamos a necestar

In [13]:
df_total=df_total.drop(['streaming_release_date'],axis=1)
df_total=df_total.drop(['runtime'],axis=1)
df_total=df_total.drop(['tomatometer_rating'],axis=1)
df_total=df_total.drop(['tomatometer_count'],axis=1)
df_total=df_total.drop(['audience_rating'],axis=1)
df_total=df_total.drop(['audience_count'],axis=1)
df_total=df_total.drop(['tomatometer_top_critics_count'],axis=1)
df_total=df_total.drop(['tomatometer_fresh_critics_count'],axis=1)
df_total=df_total.drop(['tomatometer_rotten_critics_count'],axis=1)
df_total=df_total.drop(['review_score'],axis=1)
df_total=df_total.drop(['review_date'],axis=1)



In [14]:
df_total

rotten_tomatoes_link        critic_name  top_critic  \
0                  m/0814255    Andrew L. Urban       False   
1                  m/0814255      Louise Keller       False   
2                  m/0814255                NaN       False   
3                  m/0814255       Ben McEachen       False   
4                  m/0814255        Ethan Alter        True   
...                      ...                ...         ...   
1129882          m/zulu_dawn      Chuck O'Leary       False   
1129883          m/zulu_dawn          Ken Hanke       False   
1129884          m/zulu_dawn    Dennis Schwartz       False   
1129885          m/zulu_dawn  Christopher Lloyd       False   
1129886          m/zulu_dawn     Brent McKnight       False   

                          publisher_name review_type  \
0                         Urban Cinefile       Fresh   
1                         Urban Cinefile       Fresh   
2                    FILMINK (Australia)       Fresh   
3                Sunday Mail (Australia)       Fresh   
4                     Hollywood Reporter      Rotten   
...                                  ...         ...   
1129882                 Fantastica Daily      Rotten   
1129883  Mountain Xpress (Asheville, NC)       Fresh   
1129884    Dennis Schwartz Movie Reviews       Fresh   
1129885          Sarasota Herald-Tribune      Rotten   
1129886             The Last Thing I See      Rotten   

                                            review_content  \
0        A fantasy adventure that fuses Greek mythology...   
1        Uma Thurman as Medusa, the gorgon with a coiff...   
2        With a top-notch cast and dazzling special eff...   
3        Whether audiences will get behind The Lightnin...   
4        What's really lacking in The Lightning Thief i...   
...                                                    ...   
1129882                                                NaN   
1129883  Seen today, it's not only a startling indictme...   
1129884  A rousing visual spectacle that's a prequel of...   
1129885  A simple two-act story: Prelude to war, and th...   
1129886  Rides the line between being a pure artifact o...   

                                               movie_title  \
0        Percy Jackson & the Olympians: The Lightning T...   
1        Percy Jackson & the Olympians: The Lightning T...   
2        Percy Jackson & the Olympians: The Lightning T...   
3        Percy Jackson & the Olympians: The Lightning T...   
4        Percy Jackson & the Olympians: The Lightning T...   
...                                                    ...   
1129882                                          Zulu Dawn   
1129883                                          Zulu Dawn   
1129884                                          Zulu Dawn   
1129885                                          Zulu Dawn   
1129886                                          Zulu Dawn   

                                                movie_info  \
0        Always trouble-prone, the life of teenager Per...   
1        Always trouble-prone, the life of teenager Per...   
2        Always trouble-prone, the life of teenager Per...   
3        Always trouble-prone, the life of teenager Per...   
4        Always trouble-prone, the life of teenager Per...   
...                                                    ...   
1129882  Sir Henry Bartle Frere's (John Mills) vastly o...   
1129883  Sir Henry Bartle Frere's (John Mills) vastly o...   
1129884  Sir Henry Bartle Frere's (John Mills) vastly o...   
1129885  Sir Henry Bartle Frere's (John Mills) vastly o...   
1129886  Sir Henry Bartle Frere's (John Mills) vastly o...   

                                         critics_consensus content_rating  \
0        Though it may seem like just another Harry Pot...             PG   
1        Though it may seem like just another Harry Pot...             PG   
2        Though it may seem like just another Harry Pot...             PG   
3        Though it may seem like ju

#Preprocesamiento

Iniciamos un preprocesamiento tomando en base a un archivo de stopwords, seguido instalamos punkt y punk_tab lo cual son los recursos necesarios de NLTK para tokenizar texto en oraciones y palabras

In [15]:


# Descargar recursos necesarios para NLTK
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('punkt_tab')




[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

##En base al archivo que tenemos de stopwords, cargamos para eliminar estas mismas

In [16]:

# Cargar archivo de stopwords proporcionado
with open('stopwords', 'r') as f:
    custom_stopwords = set(f.read().splitlines())


###En esta parte se define la función para el preprocesamiento

In [17]:

# Función de preprocesamiento
def preprocess_text(text):
    # 1. Convertir a minúsculas
    text = text.lower()

    # 2. Eliminar caracteres especiales
    text = re.sub(r'[^a-z\s]', '', text)

    # 3. Eliminar stopwords
    words = text.split()
    words = [word for word in words if word not in custom_stopwords]

    # 4. Lematización
    lemmatizer = WordNetLemmatizer()
    words = [lemmatizer.lemmatize(word) for word in words]

    return ' '.join(words)



###Aplicamos la función a las 3 columnas a utilizar, que son el titulo de la película, el género y el contenido review

In [19]:

# Aplicar preprocesamiento a las columnas específicas
columns_to_preprocess = ['movie_title', 'genres', 'review_content']
for column in columns_to_preprocess:
    df_total[column] = df_total[column].astype(str).apply(preprocess_text)

# Vectorización con TF-IDF
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(df_total['review_content'])  # Solo para review_content


Ahora habiendo terminado la función y el vectorizado procedesmos a realizar la función de procesar y busqeuda de querys

In [22]:

# Función para procesar consultas y buscar resultados
def search_query(query, tfidf_matrix, df_total, vectorizer):
    # Preprocesar la consulta
    processed_query = preprocess_text(query)

    # Vectorizar la consulta
    query_vector = vectorizer.transform([processed_query])

    # Calcular similitud de coseno
    similarities = cosine_similarity(query_vector, tfidf_matrix).flatten()

    # Obtener índices de las películas más relevantes
    relevant_indices = similarities.argsort()[::-1][:10]  # Top 10 resultados

    # Mostrar resultados ordenados
    results = df_total.iloc[relevant_indices]
    results['similarity'] = similarities[relevant_indices]
    return results[['movie_title', 'genres', 'similarity']]


###Por ultimo se determina la similitud que se obtuvo de la busqueda  y adicional se agrego los resultados con el porcentaje de similitud

In [23]:

# Ejemplo de consulta
query = "Películas sobre viajes espaciales"
results = search_query(query, tfidf_matrix, df_total, vectorizer)

# Mostrar resultados
print(results)
# Mostrar resultados con porcentaje de similitud
results = search_query(query, tfidf_matrix, df_total, vectorizer)
results['similarity'] = (results['similarity'] * 100).round(2)  # Convertir a porcentaje
print(results)

<ipython-input-22-cb7ea5f6d973>:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results['similarity'] = similarities[relevant_indices]


                       movie_title  \
340538                fiddler roof   
1035528                   thirteen   
551600                    man past   
370745                        real   
129413             american psycho   
1112564                     paddle   
363579                frozen river   
389228   grave firefly hotaru haka   
146845         assassination tango   
85654                                

                                                    genres  similarity  
340538   classic comedy drama kid family musical perfor...    0.308602  
1035528                                              drama    0.289632  
551600                art house international comedy drama    0.285193  
370745                            comedy drama gay lesbian    0.281453  
129413                comedy drama horror mystery suspense    0.281131  
1112564                                             comedy    0.256360  
363579                                               drama    0.251773

<ipython-input-22-cb7ea5f6d973>:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results['similarity'] = similarities[relevant_indices]


##Conclusion
El sistema de recuperación de información ha logrado identificar un conjunto de 10 películas relevantes basándose en la similitud textual con la consulta ingresada. Los resultados muestran que la película "fiddler roof" obtuvo la mayor puntuación de similitud, con un 30.86%, indicando una correspondencia significativa con los términos de la consulta. Por otro lado, la décima película, con un 24.16% de similitud, representa un nivel de relevancia más bajo pero aún relacionado con la búsqueda.

El rango de similitudes sugiere que las películas encontradas tienen distintos grados de afinidad con la consulta, lo cual es una característica típica de los sistemas basados en TF-IDF y similitud de coseno. Además, la diversidad en los géneros asociados a las películas indica que el sistema puede recuperar resultados que abarcan múltiples categorías relevantes, como drama, comedia o ciencia ficción, en función del contenido textual de las reseñas.